### PR TITLE
1678 Rename from time to timestamp in the database

### LINKED ISSUE #1678: Rename from `time` to `timestamp` in the database
To avoid the confusion of having to use `.getTime().getTime()`, replace usages of `Calendar time;` with `Calendar timestamp;`.

> [!WARNING]
> Requires DB schema migration. See [DB_MIGRATION.md](https://github.com/elimu-ai/webapp/blob/main/DB_MIGRATION.md) for details.

### PR DESCRIPTION
**Pull request by alexander-kuruvilla**
Closes #1678 

Changes:
Updated the name of objects name time to timestamp.
Updated setTime() to setTimestamp().
Updated getTime() to getTimestamp().
Updated references to getters and setters in the code.

### CODE REVIEW

File: src/main/java/ai/elimu/model/analytics/LearningEvent.java [hunk 1]
```diff
@@ -42,12 +42,12 @@ public abstract class LearningEvent extends BaseEntity {
...
 
-    public Calendar getTime() {
-        return time;
+    public Calendar getTimestamp() {
+        return timestamp;
     }
 
-    public void setTime(Calendar time) {
-        this.time = time;
+    public void setTimestamp(Calendar time) {
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/model/analytics/LearningEvent.java:49]:
```suggestion
    public void setTimestamp(Calendar timestamp) {
```

File: src/main/java/ai/elimu/model/contributor/ContributionEvent.java [hunk 2]
```diff
@@ -65,8 +65,8 @@ public void setContributor(Contributor contributor) {
         this.contributor = contributor;
     }
 
-    public void setTime(Calendar time) {
-        this.time = time;
+    public void setTimestamp(Calendar time) {
+        this.timestamp = time;
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/model/contributor/ContributionEvent.java:69]:
```suggestion
    public void setTimestamp(Calendar timestamp) {
        this.timestamp = timestamp;
```

File: src/main/java/ai/elimu/model/contributor/PeerReviewEvent.java [hunk 3]
```diff
@@ -60,11 +60,11 @@ public void setComment(String comment) {
...
 
-    public Calendar getTime() {
-        return time;
+    public Calendar getTimestamp() {
+        return timestamp;
     }
 
     public void setTime(Calendar time) {
-        this.time = time;
+        this.timestamp = time;
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/model/contributor/PeerReviewEvent.java:68]:
```suggestion
        this.timestamp = timestamp;
```

File: src/main/java/ai/elimu/rest/v2/analytics/LetterLearningEventsRestController.java [hunk 4]
```diff
@@ -120,7 +120,7 @@ public String handleUploadCsvRequest(
                 long timeInMillis = Long.valueOf(csvRecord.get("time"));
                 Calendar time = Calendar.getInstance();
                 time.setTimeInMillis(timeInMillis);
-                letterLearningEvent.setTime(time);
+                letterLearningEvent.setTimestamp(time);
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/rest/v2/analytics/LetterLearningEventsRestController.java:123]:
```suggestion
                letterLearningEvent.setTimestamp(timestamp);
```

File: src/main/java/ai/elimu/rest/v2/analytics/StoryBookLearningEventsRestController.java [hunk 5]
```diff
@@ -120,7 +120,7 @@ public String handleUploadCsvRequest(
                 long timeInMillis = Long.valueOf(csvRecord.get("time"));
                 Calendar time = Calendar.getInstance();
                 time.setTimeInMillis(timeInMillis);
-                storyBookLearningEvent.setTime(time);
+                storyBookLearningEvent.setTimestamp(time);
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/rest/v2/analytics/StoryBookLearningEventsRestController.java:123]:
```suggestion
                storyBookLearningEvent.setTimestamp(timestamp);
```

File: src/main/java/ai/elimu/rest/v2/analytics/WordLearningEventsRestController.java [hunk 6]
```diff
@@ -120,7 +120,7 @@ public String handleUploadCsvRequest(
                 long timeInMillis = Long.valueOf(csvRecord.get("time"));
                 Calendar time = Calendar.getInstance();
                 time.setTimeInMillis(timeInMillis);
-                wordLearningEvent.setTime(time);
+                wordLearningEvent.setTimestamp(time);
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/rest/v2/analytics/WordLearningEventsRestController.java:123]:
```suggestion
                wordLearningEvent.setTimestamp(timestamp);
```

File: src/main/java/ai/elimu/util/csv/CsvAnalyticsExtractionHelper.java [hunk 7]
```diff
@@ -63,7 +63,7 @@ public static List<StoryBookLearningEvent> getStoryBookLearningEventsFromCsvBack
                 long timeInMillis = Long.valueOf(csvRecord.get("time"));
                 Calendar time = Calendar.getInstance();
                 time.setTimeInMillis(timeInMillis);
-                storyBookLearningEvent.setTime(time);
+                storyBookLearningEvent.setTimestamp(time);
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/util/csv/CsvAnalyticsExtractionHelper.java:66]:
```suggestion
                storyBookLearningEvent.setTimestamp(timestamp);
```

File: src/main/resources/META-INF/jpa-schema-export.sql [hunk 8]
```diff
@@ -205,8 +205,8 @@
        id bigint not null auto_increment,
         comment varchar(1000),
         revisionNumber integer,
-        time datetime,
         timeSpentMs bigint,
+        timestamp datetime,
```
Comment (jo-elimu) [inline src/main/resources/META-INF/jpa-schema-export.sql:209]: A database migration script is missing for this file.

**Review by jo-elimu — CHANGES_REQUESTED** [review]
Thank you, @alexander-kuruvilla. Note that the webapp is already running in production, so we need to update the underlying database structure whenever we modify the JPA entities.

So your changes in `jpa-schema-export.sql` would have to be reflected in a database migration script. See more info here: https://github.com/elimu-ai/webapp/blob/main/DB_MIGRATION.md

---

FYI, these are the tables that would have to be migrated:

```sql
MariaDB [webapp-HIN]> SHOW TABLES LIKE '%Event';
+-------------------------------+
| Tables_in_webapp-HIN (%Event) |
+-------------------------------+
| AudioContributionEvent        |
| AudioPeerReviewEvent          |
| ImageContributionEvent        |
| LetterContributionEvent       |
| LetterLearningEvent           |
| LetterSoundContributionEvent  |
| LetterSoundPeerReviewEvent    |
| NumberContributionEvent       |
| NumberPeerReviewEvent         |
| SoundContributionEvent        |
| StoryBookContributionEvent    |
| StoryBookLearningEvent        |
| StoryBookPeerReviewEvent      |
| WordContributionEvent         |
| WordLearningEvent             |
| WordPeerReviewEvent           |
+-------------------------------+
16 rows in set (0.001 sec)
```

And here is an example of what a table currently looks like:

```sql
MariaDB [webapp-HIN]> DESCRIBE LetterLearningEvent;
+-------------------+--------------+------+-----+---------+----------------+
| Field             | Type         | Null | Key | Default | Extra          |
+-------------------+--------------+------+-----+---------+----------------+
| id                | bigint(20)   | NO   | PRI | NULL    | auto_increment |
| androidId         | varchar(255) | NO   |     | NULL    |                |
| time              | datetime     | NO   |     | NULL    |                |
| learningEventType | varchar(255) | NO   |     | NULL    |                |
| letterText        | varchar(255) | NO   |     | NULL    |                |
| application_id    | bigint(20)   | YES  | MUL | NULL    |                |
| letter_id         | bigint(20)   | YES  | MUL | NULL    |                |
| packageName       | varchar(255) | NO   |     | NULL    |                |
+-------------------+--------------+------+-----+---------+----------------+
8 rows in set (0.001 sec)
```

**Comment by alexander-kuruvilla** [comment]
@jo-elimu thanks for your guidance and input, I will work on those.

**Review by jo-elimu — CHANGES_REQUESTED** [review]
@alexander-kuruvilla The changes need to be reflected in the frontend files as well. For example the `WEB-INF/jsp/content/letter/edit.jsp`:

```java
Caused by: javax.el.PropertyNotFoundException: Property [time] not found on type [ai.elimu.model.contributor.LetterContributionEvent]
    at javax.el.BeanELResolver$BeanProperties.get (BeanELResolver.java:253)
    at javax.el.BeanELResolver$BeanProperties.access$300 (BeanELResolver.java:206)
    at javax.el.BeanELResolver.property (BeanELResolver.java:322)
    at javax.el.BeanELResolver.getValue (BeanELResolver.java:93)
    at org.apache.jasper.el.JasperELResolver.getValue (JasperELResolver.java:125)
    at org.apache.el.parser.AstValue.getValue (AstValue.java:168)
    at org.apache.el.ValueExpressionImpl.getValue (ValueExpressionImpl.java:189)
    at org.apache.jasper.runtime.PageContextImpl.proprietaryEvaluate (PageContextImpl.java:690)
    at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspx_meth_fmt_005fformatDate_005f0 (edit_jsp.java:1411)
    at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspx_meth_c_005fforEach_005f0 (edit_jsp.java:1305)
    at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspx_meth_content_005fsection_005f0 (edit_jsp.java:365)
    at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspService (edit_jsp.java:227)
```

<kbd></kbd> --> <kbd></kbd>

> [!TIP]
> For a quick sign-on during localhost development, use this URL: http://localhost:8080/webapp/sign-on/test/role-CONTRIBUTOR

**Comment by alexander-kuruvilla** [comment]
> @alexander-kuruvilla The changes need to be reflected in the frontend files as well. For example the `WEB-INF/jsp/content/letter/edit.jsp`:
> 
> ```java
> Caused by: javax.el.PropertyNotFoundException: Property [time] not found on type [ai.elimu.model.contributor.LetterContributionEvent]
>     at javax.el.BeanELResolver$BeanProperties.get (BeanELResolver.java:253)
>     at javax.el.BeanELResolver$BeanProperties.access$300 (BeanELResolver.java:206)
>     at javax.el.BeanELResolver.property (BeanELResolver.java:322)
>     at javax.el.BeanELResolver.getValue (BeanELResolver.java:93)
>     at org.apache.jasper.el.JasperELResolver.getValue (JasperELResolver.java:125)
>     at org.apache.el.parser.AstValue.getValue (AstValue.java:168)
>     at org.apache.el.ValueExpressionImpl.getValue (ValueExpressionImpl.java:189)
>     at org.apache.jasper.runtime.PageContextImpl.proprietaryEvaluate (PageContextImpl.java:690)
>     at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspx_meth_fmt_005fformatDate_005f0 (edit_jsp.java:1411)
>     at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspx_meth_c_005fforEach_005f0 (edit_jsp.java:1305)
>     at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspx_meth_content_005fsection_005f0 (edit_jsp.java:365)
>     at org.apache.jsp.WEB_002dINF.jsp.content.letter.edit_jsp._jspService (edit_jsp.java:227)
> ```
> 
>  --> 
> 
> Tip
> 
> For a quick sign-on during localhost development, use this URL: http://localhost:8080/webapp/sign-on/test/role-CONTRIBUTOR

Got it. I wasn't able to check out the frontend, thanks for that tip. I'll update the jsps and go over the pages. Thanks.

Also I wanted to run something by you regarding coderabbitai's query optimisation suggestion, should I make the changes along with this or does that have to be another issue?

File: src/main/java/ai/elimu/dao/jpa/LetterSoundContributionEventDaoJpa.java [hunk 9]
```diff
@@ -46,8 +46,8 @@ public List<LetterSoundContributionEvent> readMostRecentPerLetterSound() throws
         return em.createQuery(
             "SELECT e " + 
             "FROM LetterSoundContributionEvent e " +
-            "WHERE e.time IN (SELECT MAX(time) FROM LetterSoundContributionEvent GROUP BY letterSound_id) " + // TODO: replace with "NOT EXISTS"? - https://stackoverflow.com/a/25694562
-            "ORDER BY e.time ASC")
+            "WHERE e.timestamp IN (SELECT MAX(timestamp) FROM LetterSoundContributionEvent GROUP BY letterSound_id) " + // TODO: replace with "NOT EXISTS"? - https://stackoverflow.com/a/25694562
+            "ORDER BY e.timestamp ASC")
```
[the remark this answers is not in the document]
Comment (jo-elimu) [reply src/main/java/ai/elimu/dao/jpa/LetterSoundContributionEventDaoJpa.java:50]: @coderabbitai Yes, please.

**Review by jo-elimu — APPROVED** [review]
Looks ready to merge 🙂

**Pull request merged**